tests/mcmc/test_barker.py

```diff
@@ -51,5 +61,69 @@ def _test_samples_vs_pdf(samples, pdf):
     )
 
 
+class BarkerPreconditioiningTest(chex.TestCase):
+    @parameterized.parameters([1234, 5678])
+    def test_preconditioning_matrix(self, seed):
+        """Test two different ways of using pre-conditioning matrix has exactly same effect.
+
+        We follow the discussion in Appendix G of the Barker 2020 paper.
+        """
+        from blackjax.mcmc.barker import _get_mass_matrix_sqrt
+
+        key = jax.random.key(seed)
+        init_key, inference_key = jax.random.split(key, 2)
+
+        # setup some 2D multivariate normal model
+        # setup sampling mean and cov
+        true_x = jnp.array([0.0, 1.0])
+        data = jax.random.normal(init_key, shape=(1000,)) * true_x[1] + true_x[0]
+        assert data.shape == (1000,)
+
+        # some non-diagonal positive-defininte matrix for pre-conditioning
+        inv_mass_matrix = jnp.array([[1, 0.1], [0.1, 1]])
+        C_t, C_t_inv = _get_mass_matrix_sqrt(inv_mass_matrix)
+
+        # define barker kernel two ways
+        # non-scaled, use pre-conditioning
+        def logdensity(x, data):
+            mu_prior = stats.norm.logpdf(x[0], loc=0, scale=1)
+            sigma_prior = stats.uniform.logpdf(x[1], 0.0, 3.0)
+            return mu_prior + sigma_prior + jnp.sum(stats.norm.logcdf(data, x[0], x[1]))
+
+        logposterior_fn1 = functools.partial(logdensity, data=data)
+        barker1 = blackjax.barker_proposal(logposterior_fn1, 1e-1, inv_mass_matrix)
+        state1 = barker1.init(true_x)
+
+        # scaled, trivial pre-conditioning
+        def scaled_logdensity(x_scaled, data, C_t):
+            return logdensity(C_t.dot(x_scaled), data)
+
+        logposterior_fn2 = functools.partial(scaled_logdensity, data=data, C_t=C_t)
+        barker2 = blackjax.barker_proposal(logposterior_fn2, 1e-1, jnp.eye(2))
+
+        true_x_trans = C_t_inv.dot(true_x)
+        state2 = barker2.init(true_x_trans)
+
+        _, states1 = run_inference_algorithm(
+            rng_key=inference_key,
+            initial_state=state1,
+            inference_algorithm=barker1,
+            transform=lambda state, info: state.position,
+            num_steps=1000,
+        )
+
+        _, states2 = run_inference_algorithm(
+            rng_key=inference_key,
+            initial_state=state2,
+            inference_algorithm=barker2,
+            transform=lambda state, info: state.position,
+            num_steps=1000,
+        )
+
+        # states should be the exact same with same random key after transforming
+        states2_trans = C_t.dot(states2.T).T
+        np.testing.assert_allclose(states1, states2_trans, atol=1e-2)
```

Comment:
Does this pass with higher precision -> this should really hold up to numerical precision.

Reply:
I will check, you are right

Reply:
removed the tolerance in the latest commit so now it's default jax.numpy tolerance